core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -116,4 +116,13 @@ private[scheduler] abstract class Stage(
 
   /** Returns the sequence of partition ids that are missing (i.e. needs to be computed). */
   def findMissingPartitions(): Seq[Int]
+
+  def isIndeterminate: Boolean =
+    rdd.outputDeterministicLevel == DeterministicLevel.INDETERMINATE
+
+  /**
+   * Clear the intermediate status (e.g. map status, success task, etc) for the stage, currently
+   * we call the function before indeterminate stage rerunning.
+   */
+  def clearIntermediateState(): Unit = {}
```

Comment:
why it's defined in a `Stage`? These states are not held by Stage, aren't they?

Reply:
```
These states are not held by Stage, aren't they?
```
Yep, currently these states are held by map output tracker for shuffle map stage.
```
why it's defined in a Stage? 
```
We can directly call mapOutputTracker.unregisterAllMapOutput in DAGScheduler just for shuffle map stage, but I think the cleaning work is common for all kind stages. For result stage we also need do this and the states including the output result.

Reply:
All the logic has been moved in DAGScheduler, as discussed in https://github.com/apache/spark/pull/24110#discussion_r272448560.